Snowflake.Data/Core/SFSession.cs

```diff
@@ -486,22 +486,25 @@ internal void heartbeat()
                     {
                         if (response.code == SF_SESSION_EXPIRED_CODE)
                         {
-                            logger.Debug("SFSession::heartbeat session token expired and retry heartbeat");
+                            logger.Debug($"SFSession ::heartbeat Session ID: {sessionId} session token expired and retry heartbeat");
                             try
                             {
                                 renewSession();
                             }
                             catch (Exception ex)
                             {
-                                logger.Error("renew session failed.", ex);
-                                throw;
+                                // Since we don't lock the heart beat queue when sending
+                                // the heart beat, it's possible that the session get
+                                // closed when sending renew request and caused exception
+                                // thrown from renewSession(), simply ignore that
+                                logger.Error($"renew session (ID: {sessionId}) failed.", ex);

```

Comment:
Do we need to break the loop after some try?

Reply:
fixed